content/5_Plat/RURQ.mdx

```diff
@@ -221,9 +221,171 @@ queries.
+
+Given the two different update queries, a total of four different situations might take place as a result of any update. 
+
+-	Range add when $lzSet$ equals 0: Simply add the new value to the pre-existing value. 
+
+-	Range add when $lzSet$ doesn't equal 0: Add the new value to $lzSet$ and clear $lzAdd$. 
+
+-	Range set when $lzAdd$ equals 0: Simply update the $lzSet$ value. 
+
+-	Range set when $lzAdd$ doesn't equal 0: Again simply update the $lzSet$ value since a set update will override all previous add updates. 
+
+Given the mechanics behind the ```pushdown``` function, all you need is a regular range sum segment tree to complete the question. 
+
+```cpp
+
+#include <iostream>
+using ll = long long;
+using namespace std;
+ 
+const int maxN = 2e5+5;
+ 
+int N,Q;
+int a[maxN];
+ 
+struct node{
+	ll val;
+	int l,r;
+	ll lzAdd; 
+	ll lzSet;
+	node(){};
+}tree[maxN<<2];
+ 
+#define lc p<<1
+#define rc (p<<1)+1
+ 
+inline void pushup(int p){
+	tree[p].val = tree[lc].val + tree[rc].val;
+	return;
+}
+ 
+void pushdown(int p){
+	//leaf node
+	if(tree[p].l == tree[p].r)return;
+	//lazy: range set
+	if(tree[p].lzSet != 0){
+		tree[lc].lzSet = tree[rc].lzSet = tree[p].lzSet;
+		tree[lc].val = (tree[lc].r - tree[lc].l + 1) * tree[p].lzSet;
+		tree[rc].val = (tree[rc].r - tree[rc].l + 1) * tree[p].lzSet;
+		tree[lc].lzAdd = tree[rc].lzAdd = 0;
+		tree[p].lzSet = 0;
+	}
+	else if(tree[p].lzAdd != 0){ //lazy: range add
+		if(tree[lc].lzSet == 0)tree[lc].lzAdd += tree[p].lzAdd;
+		else {
+			tree[lc].lzSet += tree[p].lzAdd;
+			tree[lc].lzAdd = 0;
+		}
+		if(tree[rc].lzSet == 0)tree[rc].lzAdd += tree[p].lzAdd;
+		else{
+			tree[rc].lzSet += tree[p].lzAdd;
+			tree[rc].lzAdd = 0;
+		}
+		tree[lc].val += (tree[lc].r - tree[lc].l + 1) * tree[p].lzAdd;
+		tree[rc].val += (tree[rc].r - tree[rc].l + 1) * tree[p].lzAdd;
+		tree[p].lzAdd = 0;
+	}
+	return;
+}
+ 
+void build(int p, int l, int r){
+	tree[p].l = l;
+	tree[p].r = r;
+	tree[p].lzAdd = tree[p].lzSet = 0;
+	if(l == r){
+		tree[p].val = a[l];
+		return;
+	}
+	int mid = (l+r)>>1;
+	build(lc,l,mid);
+	build(rc,mid+1,r);
+	pushup(p);
+	return;
+}
+ 
+void add(int p, int l, int r, int a, int b, int val){
+	if(a > r || b < l)return;
+	if(a <= l && r <= b){
+		tree[p].val += (tree[p].r - tree[p].l + 1) * val;
+		if(tree[p].lzSet == 0)tree[p].lzAdd += val;
+		else tree[p].lzSet += val;
+		return;
+	}
+	pushdown(p);
+	int mid = (l+r)>>1;
+	add(lc,l,mid,a,b,val);
+	add(rc,mid+1,r,a,b,val);
+	pushup(p);
+	return;
+}
+ 
+void set(int p, int l, int r, int a, int b, int val){
+	if(a > r || b < l)return;
+	if(a <= l && r <= b){
+		tree[p].val = (tree[p].r - tree[p].l + 1) * val;
+		tree[p].lzAdd = 0;
+		tree[p].lzSet = val;
+		return;
+	}
+	pushdown(p);
+	int mid = (l+r)>>1;
+	set(lc,l,mid,a,b,val);
+	set(rc,mid+1,r,a,b,val);
+	pushup(p);
+	return;
+}
+ 
+ll query(int p, int l, int r, int a, int b){
```

Comment:
since you store `l` and `r` in `tree[p]`, you don't need to pass `l` and `r` as arguments to this function

Reply:
True. Should I change this or is this just for future reference ?

Reply:
change

Reply:
I removed ```l``` and ```r``` in each node of the tree to minimize unnecessary redundancy.